File: python/src/n0110_balanced_binary_tree.py
from typing import Optional

from utils.tree import TreeNode


def is_balanced(root: Optional[TreeNode]) -> bool:
    return check_balance(root) is not None


def check_balance(root: Optional[TreeNode]) -> Optional[int]:
    if not root:
        return 0
    left = check_balance(root.left)
    if left is None:
        return None
    right = check_balance(root.right)
    if right is None or abs(left - right) > 1:
        return None
    return 1 + max(left, right)

File: typescript/src/n0110_balanced_binary_tree.ts
import type { TreeNode } from './utils/tree';

export function isBalanced(root: TreeNode | null): boolean {
  return checkBalance(root) !== null;
}

function checkBalance(node: TreeNode | null): number | null {
  if (!node) {
    return 0;
  }
  const left = checkBalance(node.left);
  if (left === null) {
    return null;
  }
  const right = checkBalance(node.right);
  if (right === null || Math.abs(left - right) > 1) {
    return null;
  }
  return 1 + Math.max(left, right);
}
